### ISSUE TITLE
Extend charm to resolve subgrid positions and velocities

### ISSUE THREAD

**Issue by maho3** [maho3/ltu-gobig#17]

**Comment by shivampcosmo** [comment]
<img width="572" alt="Screenshot 2024-07-30 at 11 03 18 AM" src="https://github.com/user-attachments/assets/5a98d9aa-693f-475d-8ba7-919767979a12">
<img width="586" alt="Screenshot 2024-07-30 at 11 03 24 AM" src="https://github.com/user-attachments/assets/d1517348-b60a-46a6-bcfc-e0d91e43fe11">
<img width="582" alt="Screenshot 2024-07-30 at 11 03 29 AM" src="https://github.com/user-attachments/assets/6625ad29-ad73-49a0-afb9-1a109a7996ac">
<img width="592" alt="Screenshot 2024-07-30 at 11 05 53 AM" src="https://github.com/user-attachments/assets/9e3f7814-02bc-4b0d-8d17-f33a6708e2af">

experimenting with velocities in CHARM, by implementing a separate autoregressive network that predicts the halo velocity. This is first test that assumes halo position and masses are true, and just learns the correction for velocity on top of the interpolated velocity from FASTPM field.

**Comment by tlmakinen** [comment]
how many transformations / descriptive is the velocity autoregressive network ? is it the same architecture as is used for the positions ?

**Comment by shivampcosmo** [comment]
Yes, more or less. It is stack of two normalizing flows going from Gaussian to velocity distribution (difference between FastPM and true).

**Comment by shivampcosmo** [comment]
first attempt at generalizing for cosmologies. Seems to work much better than getting the velocities from FastPM, but currently only ~10% level. 

![Image](https://github.com/user-attachments/assets/f516f6c3-081f-4634-9458-8cb069ec9bc4)
![Image](https://github.com/user-attachments/assets/2a6620ab-c0b4-4c3f-be02-685aa78afc0c)
![Image](https://github.com/user-attachments/assets/dc7b476d-d797-46a2-80be-a2462c6fe9cc)
![Image](https://github.com/user-attachments/assets/0d40fdc2-d1f1-4df7-acd4-892e7b7a805c)

**Comment by shivampcosmo** [comment]
CHARM updated training for both masses and velocities. Now the network goes to smaller halo masses (M > 5e12 Msun/h) and also trains much better on multiple GPUs. Also the code is stand-alone installable and pushed here : https://github.com/shivampcosmo/CHARM . Integration with current ltu-cmass pipeline in progress. 

![Image](https://github.com/user-attachments/assets/2a1ee7a5-8671-4ad0-9b92-04c7b6ba2f7d)
![Image](https://github.com/user-attachments/assets/5fb153ae-bda6-4c40-867f-e7d057f0c85c)
![Image](https://github.com/user-attachments/assets/0e55870f-70f8-48c9-b7de-81fb28fd9d1c)
![Image](https://github.com/user-attachments/assets/a29ff124-5e41-4d6f-9503-866bf4301729)

**Comment by tlmakinen** [comment]
some updated objectives after conversations with @shivampcosmo @maho3:

- look at modelling positions on nodes within $$8^3$$ voxels using a cheap fully-connected GNN layer. 
- TODO: look at (true) displacement distributions from COM in each voxel
- heavy-tailed studentT MDN might be the best parameterization for the loss $$- \ln p(x, \hat{x}) $$

**Comment by tlmakinen** [comment]
Updates from LtU meeting:

Predicting positions within voxels might be best done as a residual learning problem. Looking at the difference in position between the heaviest halo $x_0$ and the next heaviest halo $x_1$ (and so on), we see a zero-centered heavy-tailed distribution. 
 ![Image](https://github.com/user-attachments/assets/f6531589-c916-428b-a225-50d75f212096)

Would be good to model this with an MDN neural studentT e.g. following https://arxiv.org/pdf/2006.06599
